# User Engagement Trends in Dognition

by Michelle Aguiar

## Overview
This case study uses SQL to analyze user engagement patterns within Dognition, a cognitive testing platform for dogs. The objective was to uncover when users are most active and identify the geographic regions with the
highest engagement. Understanding these trends supports strategic decisions around communication timing, user segmentation, and regional outreach.

The analysis utilized data from the `complete_tests`, `dogs`, and `users` tables. 

**Key techniques included:**
* Data cleaning and filtering to remove flagged users/dogs
* Aggregation to track engagement by weekday and year
* Timezone adjustments for accuracy in behavioral timing
* Geographic segmentation to identify top markets

In [1]:
%load_ext sql
%sql mysql://studentuser:studentpw@localhost/dognitiondb
%sql USE dognitiondb

 * mysql://studentuser:***@localhost/dognitiondb
0 rows affected.


[]

## Engagement Peaks by Day of the Week

The analysis first explored test activity by day of the week to uncover behavioral trends. Using the `DAYOFWEEK()` function, weekday values were extracted, labeled, and aggregated to quantify completions by day. 

Results revealed **Sunday as the highest engagement day (~33,190 completions)** and Friday as the lowest.

In [2]:
%%sql

# Extract weekday from timestamp
SELECT 
    created_at, 
    DAYOFWEEK(created_at) AS weekday_num
FROM complete_tests
LIMIT 49200;

# Add weekday labels using CASE
SELECT 
    created_at, 
    DAYOFWEEK(created_at) AS weekday_num,
    CASE
        WHEN DAYOFWEEK(created_at) = 1 THEN 'Su'
        WHEN DAYOFWEEK(created_at) = 2 THEN 'Mo'
        WHEN DAYOFWEEK(created_at) = 3 THEN 'Tu'
        WHEN DAYOFWEEK(created_at) = 4 THEN 'We'
        WHEN DAYOFWEEK(created_at) = 5 THEN 'Th'
        WHEN DAYOFWEEK(created_at) = 6 THEN 'Fr'
        WHEN DAYOFWEEK(created_at) = 7 THEN 'Sa'
    END AS daylabel
FROM complete_tests
LIMIT 49200;

# Aggregate test counts by daylabel
SELECT 
    DAYOFWEEK(created_at) AS weekday_num,
    COUNT(created_at) AS numtests,
    CASE
        WHEN DAYOFWEEK(created_at) = 1 THEN 'Su'
        WHEN DAYOFWEEK(created_at) = 2 THEN 'Mo'
        WHEN DAYOFWEEK(created_at) = 3 THEN 'Tu'
        WHEN DAYOFWEEK(created_at) = 4 THEN 'We'
        WHEN DAYOFWEEK(created_at) = 5 THEN 'Th'
        WHEN DAYOFWEEK(created_at) = 6 THEN 'Fr'
        WHEN DAYOFWEEK(created_at) = 7 THEN 'Sa'
    END AS daylabel
FROM complete_tests
GROUP BY daylabel
ORDER BY numtests DESC;

 * mysql://studentuser:***@localhost/dognitiondb
49200 rows affected.
49200 rows affected.
7 rows affected.


[(1, 33190, 'Su'),
 (2, 30195, 'Mo'),
 (3, 27989, 'Tu'),
 (7, 27899, 'Sa'),
 (4, 26473, 'We'),
 (5, 24420, 'Th'),
 (6, 23080, 'Fr')]

Sunday had the highest number of test completions (~33,190), while Friday had the lowest.

## Data Cleansing to Improve Accuracy

To ensure data quality and relevance, records linked to users or dogs marked for exclusion were removed. After cleansing, only 34,121 out of 950,331 records met the quality threshold, creating a reliable baseline for subsequent analysis.

In [3]:
%%sql

# Total number of dogs linked to users before applying exclusion filters
SELECT 
    COUNT(*) AS total_linked_dogs
FROM dogs d
INNER JOIN users u ON d.user_guid = u.user_guid;

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(950331,)]

In [4]:
%%sql

# Count of unique dog_guid entries before cleaning
SELECT 
    COUNT(DISTINCT dog_guid) AS unique_dogs
FROM dogs d
JOIN users u ON d.user_guid = u.user_guid;

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(35048,)]

In [5]:
%%sql

# Count of valid unique dog_guid values (excluding flagged users and dogs)
SELECT 
    COUNT(DISTINCT dog_guid) AS valid_unique_dogs
FROM dogs d
JOIN users u ON d.user_guid = u.user_guid
WHERE (u.exclude IS NULL OR u.exclude = 0)
  AND (d.exclude IS NULL OR d.exclude = 0);

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(34121,)]

Of the 950,331 dogs initially linked to users, 35,048 were unique, and 34,121 met the criteria for clean analysis. This refinement ensured that only unflagged and valid records were retained, improving the accuracy of downstream insights.

## Year-Over-Year Consistency in Engagement

The report evaluated whether weekly engagement trends remained stable across years. Cleaned test data was grouped by weekday and year, confirming that Sunday and Monday consistently showed the highest activity from 2013 to 2015.

In [6]:
%%sql

# Group cleaned test data by year and weekday to detect recurring patterns
SELECT 
    DAYOFWEEK(c.created_at) AS dayasnum, 
    YEAR(c.created_at) AS year, 
    COUNT(c.created_at) AS numtests,
    CASE 
        WHEN DAYOFWEEK(c.created_at) = 1 THEN 'Su'
        WHEN DAYOFWEEK(c.created_at) = 2 THEN 'Mo'
        WHEN DAYOFWEEK(c.created_at) = 3 THEN 'Tu'
        WHEN DAYOFWEEK(c.created_at) = 4 THEN 'We'
        WHEN DAYOFWEEK(c.created_at) = 5 THEN 'Th'
        WHEN DAYOFWEEK(c.created_at) = 6 THEN 'Fr'
        WHEN DAYOFWEEK(c.created_at) = 7 THEN 'Sa'
    END AS daylabel
FROM complete_tests c
JOIN (
    SELECT DISTINCT dog_guid
    FROM dogs d
    JOIN users u ON d.user_guid = u.user_guid
    WHERE (u.exclude IS NULL OR u.exclude = 0)
      AND (d.exclude IS NULL OR d.exclude = 0)
) AS dogs_cleaned ON c.dog_guid = dogs_cleaned.dog_guid
GROUP BY year, daylabel
ORDER BY year ASC, numtests DESC;

 * mysql://studentuser:***@localhost/dognitiondb
21 rows affected.


[(1, 2013, 8203, 'Su'),
 (7, 2013, 6854, 'Sa'),
 (2, 2013, 5740, 'Mo'),
 (4, 2013, 5665, 'We'),
 (3, 2013, 5393, 'Tu'),
 (6, 2013, 4997, 'Fr'),
 (5, 2013, 4961, 'Th'),
 (2, 2014, 9309, 'Mo'),
 (1, 2014, 9210, 'Su'),
 (3, 2014, 9177, 'Tu'),
 (4, 2014, 8857, 'We'),
 (7, 2014, 8257, 'Sa'),
 (5, 2014, 7286, 'Th'),
 (6, 2014, 6475, 'Fr'),
 (1, 2015, 13623, 'Su'),
 (2, 2015, 13089, 'Mo'),
 (3, 2015, 11126, 'Tu'),
 (7, 2015, 11038, 'Sa'),
 (5, 2015, 10076, 'Th'),
 (4, 2015, 9911, 'We'),
 (6, 2015, 9555, 'Fr')]

Consistent peaks on Sunday and Monday across 2013–2015.

## Adjusting Timestamps to Reflect Local Behavior

Because timestamps were recorded in UTC, a six-hour adjustment was applied to align activity to Central Time (UTC–6). This correction offered a more accurate reflection of user-local behavior. 

Alaska and Hawaii were excluded to maintain geographic consistency. 

The Sunday and Monday pattern remained consistent post-adjustment.

In [7]:
%%sql

# Convert UTC timestamps to Central Time (UTC-6)
SELECT 
    created_at, 
    DATE_SUB(created_at, INTERVAL 6 HOUR) AS corrected_time
FROM complete_tests
LIMIT 100;

 * mysql://studentuser:***@localhost/dognitiondb
100 rows affected.


[(datetime.datetime(2013, 2, 5, 18, 26, 54), datetime.datetime(2013, 2, 5, 12, 26, 54)),
 (datetime.datetime(2013, 2, 5, 18, 31, 3), datetime.datetime(2013, 2, 5, 12, 31, 3)),
 (datetime.datetime(2013, 2, 5, 18, 32, 4), datetime.datetime(2013, 2, 5, 12, 32, 4)),
 (datetime.datetime(2013, 2, 5, 18, 32, 25), datetime.datetime(2013, 2, 5, 12, 32, 25)),
 (datetime.datetime(2013, 2, 5, 18, 32, 56), datetime.datetime(2013, 2, 5, 12, 32, 56)),
 (datetime.datetime(2013, 2, 5, 18, 33, 15), datetime.datetime(2013, 2, 5, 12, 33, 15)),
 (datetime.datetime(2013, 2, 5, 18, 33, 33), datetime.datetime(2013, 2, 5, 12, 33, 33)),
 (datetime.datetime(2013, 2, 5, 18, 33, 59), datetime.datetime(2013, 2, 5, 12, 33, 59)),
 (datetime.datetime(2013, 2, 5, 18, 34, 25), datetime.datetime(2013, 2, 5, 12, 34, 25)),
 (datetime.datetime(2013, 2, 5, 18, 34, 39), datetime.datetime(2013, 2, 5, 12, 34, 39)),
 (datetime.datetime(2013, 2, 5, 18, 34, 46), datetime.datetime(2013, 2, 5, 12, 34, 46)),
 (datetime.datetime(2013, 2, 5, 18, 35, 18), datetime.datetime(2013, 2, 5, 12, 35, 18)),
 (datetime.datetime(2013, 2, 5, 18, 35, 47), datetime.datetime(2013, 2, 5, 12, 35, 47)),
 (datetime.datetime(2013, 2, 5, 18, 36, 28), datetime.datetime(2013, 2, 5, 12, 36, 28)),
 (datetime.datetime(2013, 2, 5, 18, 36, 44), datetime.datetime(2013, 2, 5, 12, 36, 44)),
 (datetime.datetime(2013, 2, 5, 18, 57, 5), datetime.datetime(2013, 2, 5, 12, 57, 5)),
 (datetime.datetime(2013, 2, 5, 19, 1, 31), datetime.datetime(2013, 2, 5, 13, 1, 31)),
 (datetime.datetime(2013, 2, 5, 19, 4, 42), datetime.datetime(2013, 2, 5, 13, 4, 42)),
 (datetime.datetime(2013, 2, 5, 19, 7, 39), datetime.datetime(2013, 2, 5, 13, 7, 39)),
 (datetime.datetime(2013, 2, 5, 19, 15, 1), datetime.datetime(2013, 2, 5, 13, 15, 1)),
 (datetime.datetime(2013, 2, 5, 19, 21, 15), datetime.datetime(2013, 2, 5, 13, 21, 15)),
 (datetime.datetime(2013, 2, 5, 19, 28, 27), datetime.datetime(2013, 2, 5, 13, 28, 27)),
 (datetime.datetime(2013, 2, 5, 19, 51, 57), datetime.datetime(2013, 2, 5, 13, 51, 57)),
 (datetime.datetime(2013, 2, 5, 19, 58, 6), datetime.datetime(2013, 2, 5, 13, 58, 6)),
 (datetime.datetime(2013, 2, 5, 20, 5, 57), datetime.datetime(2013, 2, 5, 14, 5, 57)),
 (datetime.datetime(2013, 2, 5, 20, 8, 2), datetime.datetime(2013, 2, 5, 14, 8, 2)),
 (datetime.datetime(2013, 2, 5, 20, 35, 47), datetime.datetime(2013, 2, 5, 14, 35, 47)),
 (datetime.datetime(2013, 2, 5, 20, 47), datetime.datetime(2013, 2, 5, 14, 47)),
 (datetime.datetime(2013, 2, 5, 20, 50, 45), datetime.datetime(2013, 2, 5, 14, 50, 45)),
 (datetime.datetime(2013, 2, 5, 20, 53, 59), datetime.datetime(2013, 2, 5, 14, 53, 59)),
 (datetime.datetime(2013, 2, 5, 20, 54, 19), datetime.datetime(2013, 2, 5, 14, 54, 19)),
 (datetime.datetime(2013, 2, 5, 20, 54, 36), datetime.datetime(2013, 2, 5, 14, 54, 36)),
 (datetime.datetime(2013, 2, 5, 20, 54, 52), datetime.datetime(2013, 2, 5, 14, 54, 52)),
 (datetime.datetime(2013, 2, 5, 20, 55, 3), datetime.datetime(2013, 2, 5, 14, 55, 3)),
 (datetime.datetime(2013, 2, 5, 20, 55, 12), datetime.datetime(2013, 2, 5, 14, 55, 12)),
 (datetime.datetime(2013, 2, 5, 20, 55, 22), datetime.datetime(2013, 2, 5, 14, 55, 22)),
 (datetime.datetime(2013, 2, 5, 21, 12, 31), datetime.datetime(2013, 2, 5, 15, 12, 31)),
 (datetime.datetime(2013, 2, 5, 21, 20, 10), datetime.datetime(2013, 2, 5, 15, 20, 10)),
 (datetime.datetime(2013, 2, 5, 21, 26, 51), datetime.datetime(2013, 2, 5, 15, 26, 51)),
 (datetime.datetime(2013, 2, 5, 21, 33, 24), datetime.datetime(2013, 2, 5, 15, 33, 24)),
 (datetime.datetime(2013, 2, 5, 21, 44, 38), datetime.datetime(2013, 2, 5, 15, 44, 38)),
 (datetime.datetime(2013, 2, 5, 21, 45, 12), datetime.datetime(2013, 2, 5, 15, 45, 12)),
 (datetime.datetime(2013, 2, 5, 21, 48, 58), datetime.datetime(2013, 2, 5, 15, 48, 58)),
 (datetime.datetime(2013, 2, 5, 21, 51, 17), datetime.datetime(2013, 2, 5, 15, 51, 17)),
 (datetime.datetime(2013, 2, 5, 21, 54, 53), datetime.datetime(2013, 2, 5, 15, 54, 53)),
 (datetime.datetime(2013, 2

In [8]:
%%sql

# Time-corrected engagement by weekday and year
SELECT 
    DAYOFWEEK(DATE_SUB(c.created_at, INTERVAL 6 HOUR)) AS dayasnum, 
    YEAR(c.created_at) AS year, 
    COUNT(c.created_at) AS numtests,
    CASE
        WHEN DAYOFWEEK(DATE_SUB(c.created_at, INTERVAL 6 HOUR)) = 1 THEN 'Su'
        WHEN DAYOFWEEK(DATE_SUB(c.created_at, INTERVAL 6 HOUR)) = 2 THEN 'Mo'
        WHEN DAYOFWEEK(DATE_SUB(c.created_at, INTERVAL 6 HOUR)) = 3 THEN 'Tu'
        WHEN DAYOFWEEK(DATE_SUB(c.created_at, INTERVAL 6 HOUR)) = 4 THEN 'We'
        WHEN DAYOFWEEK(DATE_SUB(c.created_at, INTERVAL 6 HOUR)) = 5 THEN 'Th'
        WHEN DAYOFWEEK(DATE_SUB(c.created_at, INTERVAL 6 HOUR)) = 6 THEN 'Fr'
        WHEN DAYOFWEEK(DATE_SUB(c.created_at, INTERVAL 6 HOUR)) = 7 THEN 'Sa'
    END AS daylabel
FROM complete_tests c
JOIN (
    SELECT DISTINCT dog_guid 
    FROM dogs d 
    JOIN users u ON d.user_guid = u.user_guid
    WHERE (u.exclude IS NULL OR u.exclude = 0)
      AND u.country = 'US'
      AND u.state NOT IN ('HI', 'AK')
      AND (d.exclude IS NULL OR d.exclude = 0)
) AS dogs_cleaned ON c.dog_guid = dogs_cleaned.dog_guid
GROUP BY year, daylabel
ORDER BY year ASC, FIELD(daylabel, 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa', 'Su');

 * mysql://studentuser:***@localhost/dognitiondb
21 rows affected.


[(2, 2013, 3798, 'Mo'),
 (3, 2013, 3276, 'Tu'),
 (4, 2013, 3410, 'We'),
 (5, 2013, 3079, 'Th'),
 (6, 2013, 3049, 'Fr'),
 (7, 2013, 4754, 'Sa'),
 (1, 2013, 6061, 'Su'),
 (2, 2014, 7908, 'Mo'),
 (3, 2014, 6513, 'Tu'),
 (4, 2014, 5772, 'We'),
 (5, 2014, 4800, 'Th'),
 (6, 2014, 4339, 'Fr'),
 (7, 2014, 6081, 'Sa'),
 (1, 2014, 7736, 'Su'),
 (2, 2015, 8229, 'Mo'),
 (3, 2015, 6673, 'Tu'),
 (4, 2015, 6266, 'We'),
 (5, 2015, 5881, 'Th'),
 (6, 2015, 5275, 'Fr'),
 (7, 2015, 7154, 'Sa'),
 (1, 2015, 10406, 'Su')]

The Sunday/Monday engagement pattern holds even after time-zone adjustment.

## Identifying Top Performing Regions

The final phase segmented engagement by U.S. state and country to identify top-performing regions. California, New York, and Texas led U.S. activity. Internationally, the U.S., Canada, and the U.K. were the most engaged markets, suggesting opportunities for targeted outreach in English-speaking regions.

In [9]:
%%sql

# Top 5 U.S. states by unique users
SELECT 
    dogs_cleaned.state AS state, 
    COUNT(DISTINCT dogs_cleaned.user_guid) AS numusers
FROM complete_tests c
JOIN (
    SELECT DISTINCT dog_guid, u.user_guid, u.state
    FROM dogs d
    JOIN users u ON d.user_guid = u.user_guid
    WHERE (u.exclude IS NULL OR u.exclude = 0)
      AND u.country = 'US'
      AND (d.exclude IS NULL OR d.exclude = 0)
) AS dogs_cleaned ON c.dog_guid = dogs_cleaned.dog_guid
GROUP BY state
ORDER BY numusers DESC
LIMIT 5;

 * mysql://studentuser:***@localhost/dognitiondb
5 rows affected.


[('CA', 1363), ('NY', 628), ('TX', 536), ('FL', 502), ('NC', 467)]

In [10]:
%%sql

# Top 10 countries by engagement
SELECT dogs_cleaned.country AS country, COUNT(DISTINCT dogs_cleaned.user_guid) AS numusers FROM complete_tests c 
JOIN(
    SELECT DISTINCT dog_guid, u.user_guid, u.country 
    FROM dogs d JOIN users u ON d.user_guid=u.user_guid
    WHERE ((u.exclude IS NULL OR u.exclude=0)
           AND (d.exclude IS NULL OR d.exclude=0))) AS dogs_cleaned ON c.dog_guid=dogs_cleaned.dog_guid
GROUP BY country
ORDER BY numusers DESC 
LIMIT 10;

 * mysql://studentuser:***@localhost/dognitiondb
10 rows affected.


[('US', 8936),
 ('N/A', 5466),
 ('CA', 484),
 ('AU', 142),
 ('GB', 123),
 ('DE', 40),
 ('NZ', 38),
 ('DK', 34),
 ('NO', 30),
 ('FR', 23)]

California and New York lead U.S. engagement. Internationally, the U.S., Canada, and U.K. are top markets.

## Key Insights

* User activity consistently peaks on Sundays and Mondays.
* Friday consistently shows the lowest engagement.
* Weekly usage patterns are stable across years and time zones.
* High engagement is concentrated in California, New York, and English-speaking countries.

## Recommendations

To capitalize on peak engagement periods, it is recommended that Dognition schedule communications and campaigns for **Sundays** and **Mondays**, when user activity is at its highest. System maintenance and downtime should be carefully avoided on weekends, particularly Sundays, to prevent service disruptions during high-traffic periods.

Marketing resources should be concentrated in the platform’s top-performing regions: California, New York, and English-speaking countries such as the United States, Canada, the United Kingdom, and Australia. These areas consistently show higher levels of user engagement and present opportunities for deeper outreach.

Continuous monitoring of usage trends is advised to ensure alignment with user behavior over time. Lastly, geographic areas with low engagement should be investigated further, as they may reveal untapped markets or signal barriers to access that can be addressed strategically.